fdbserver/DataDistribution.actor.cpp

```diff
@@ -396,6 +424,198 @@ struct DataDistributor : NonCopyable, ReferenceCounted<DataDistributor> {
+			          [](AuditStorageState a, AuditStorageState b) {
+				          return a.id < b.id; // Inplacement sort in ascending order
+			          });
+			// Cleanup audit metadata for Failed/Complete audits
+			// Resume RUNNING audits
+			// Keep Error audits persistent
+			for (const auto& auditState : restoredAudits[auditType]) {
+				TraceEvent(SevVerbose, "AuditStorageResumeCheck", self->ddId)
+				    .detail("AuditID", auditState.id)
+				    .detail("AuditType", auditState.getType())
+				    .detail("AuditRange", auditState.range);
+				if (auditState.getPhase() == AuditPhase::Error) {
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Failed) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear both audit metadata and corresponding progress metadata
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/true));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				} else if (auditState.getPhase() == AuditPhase::Complete) {
+					if (numFinishAuditsCleared < numFinishAuditsToClear) {
+						// Clear audit metadata only
+						// No need to clear the corresponding progress metadata
+						// since it has been cleared for Complete audits
+						self->addActor.send(clearAuditMetadata(self->txnProcessor->context(),
+						                                       auditState.getType(),
+						                                       auditState.id,
+						                                       /*clearProgressMetadata=*/false));
+						numFinishAuditsCleared++;
+					}
+					continue;
+				}
+				ASSERT(auditState.getPhase() == AuditPhase::Running);
+				if (self->audits.contains(auditState.getType()) &&
+				    self->audits[auditState.getType()].contains(auditState.id)) {
+					// It is possible that the current DD is running this audit
+					// Suppose DDinit re-runs right after a new audit is persisted
+					// For this case, auditResume sees the new audit and resumes it
+					// At this point, the new audit is in the audit map
+					// Meanwhile, existing DDAuditStorage retries (since DD does not restart)
+					// Then, existing DDAuditStorageStart is conflict with auditResume
+					// auditResume cannot see the new audit in the map
+					continue;
+				}
+				TraceEvent(SevDebug, "AuditStorageResume", self->ddId)
+				    .detail("AuditID", auditState.id)
+				    .detail("AuditType", auditState.getType())
+				    .detail("AuditState", auditState.toString())
+				    .detail("NumFinishAuditsCleared", numFinishAuditsCleared)
+				    .detail("IsReady", self->auditInitialized.getFuture().isReady());
+				runAuditStorage(self, auditState, 0, "ResumeAudit");
+			}
+		}
+
+		self->auditInitialized.send(Void());
+		TraceEvent(SevDebug, "AuditStorageResumeDone", self->ddId);
```

Comment:
Can we reuse the logic in other audit actors? e.g., we load all on-disk audit, and then call a variation of `launchAudit()` for each of them?

Reply:
Yes. Currently this part reuses runAuditStorage() to start audit. launchAudit() is only used for serving audit requests.